### PR TITLE
feat(NumberInput): add status validate icons

### LINKED ISSUE #7730: Number input - add new Status variant
Add a new variant of the Number input component to support inserting a status icon into the field (potentially used to warn users about exceeding threshold limits, etc).

Core PR: https://github.com/patternfly/patternfly/pull/4983

### PR DESCRIPTION
**Pull request by gitdallas**
**What**: Closes #7730 

used `validated` property based on TextInput using that property for such styling.

example: https://patternfly-react-pr-7806.surge.sh/components/number-input#with-status

**Additional issues**:

### CODE REVIEW

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 1]
```diff
@@ -122,10 +131,15 @@ export const NumberInput: React.FunctionComponent<NumberInputProps> = ({
           </span>
         </Button>
         <input
```
Comment (mcoker) [inline packages/react-core/src/components/NumberInput/NumberInput.tsx:133]: Would it be possible to just use the `<TextInput>` component here instead of recreating this functionality?

**Review by mcoker — COMMENTED** [review]
LGTM! Left a comment about the possibility of using `<TextInput>` if we can for the number input? 

Also the validated example input is `readonly` - should we add an `onChange` handler so it matches the rest of the examples? I imagine the use case for this is the same as the default examples, except with a visual cue for when the value is within threholds.

File: packages/react-core/src/components/NumberInput/NumberInput.tsx (hunk 1 above)
Comment (gitdallas) [reply packages/react-core/src/components/NumberInput/NumberInput.tsx:133]: It's looking like this should work fine... but:

- Without changing TextInput, snapshots will need to be updated to include `aria-invalid="false"` (previously `aria-invalid` would only come in if `true`) and a few `data-ouia` properties (component-id, component-type, and safe). Should I assume this is okay and just update the snaps?
- The NumberInput component `onChange` event only has `event` for a parameter while the TextInput component has `value, event`. To use TextInput in NumberInput without a breaking change.. I'm doing the following for the TextInput prop within NumberInput, is this okay?
`{...(onChange && { onChange: (v, e) => onChange(e) })}`

File: packages/react-core/src/components/NumberInput/__tests__/__snapshots__/NumberInput.test.tsx.snap [hunk 2]
```diff
@@ -187,17 +193,21 @@ exports[`numberInput passes button props successfully 1`] = `
         </span>
       </button>
       <input
+        aria-describedby="minus-id"
```
Comment (gitdallas) [inline packages/react-core/src/components/NumberInput/__tests__/__snapshots__/NumberInput.test.tsx.snap:196]: gonna look into this...

**Review by mcoker — APPROVED** [review]
LGTM! The only thing I noticed is if I try to enter a value in that input manually and hit backspace to enter a new number, it replaces the value with `0`. That seems to just be an example thing in the `onChange` callback - up/down arrows work to change the number, so works for me!

**Comment by gitdallas** [comment]
> LGTM! The only thing I noticed is if I try to enter a value in that input manually and hit backspace to enter a new number, it replaces the value with `0`. That seems to just be an example thing in the `onChange` callback - up/down arrows work to change the number, so works for me!

@mcoker  - You'll see the same thing here https://www.patternfly.org/v4/components/number-input#with-unit-and-thresholds

It has to do with the fact we are limiting it to numbers (empty string isn't a number) between 0 and 10 in these cases.

**Comment by gitdallas** [comment]
@mcarrano - in this PR i've utilized the textinput inside of the numberinput... however, there is logic within InputGroup is checking whether any (non select/textarea/textinput) children have an ID and then using that ID to set a describedby on the TextInput

This makes sense if we are using a label with an ID.. but in this case, if we put an ID on the up/down buttons for the NumberInput it will use that ID in the describedby for the TextInput which doesn't really make sense. I can't imagine why a button would be used as a describedby.

I've made it so that Button is one of the excluded types to get an ID from for describedby, but noticed there was actually a test wanting exactly that (a button ID used for describedby) https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/InputGroup/__tests__/InputGroup.test.tsx#L16

I've updated the test to use a label instead. I've asked around a bit and it seems to be going back to you as the person who can make the decision on whether or not we can stop using a button ID in the inputGroup as the describedby for the input.

**Comment by mcoker** [comment]
> there is logic within InputGroup is checking whether any (non select/textarea/textinput) children have an ID and then using that ID to set a describedby on the TextInput

@gitdallas IMO we should have a prop that disables that, and I'd argue making that the default when we can introduce that breaking change. The core examples show that same basic `id`/`aria-describedby` association (also using `<button>`s) - https://www.patternfly.org/v4/components/input-group/html#variations. I'm guessing when we [created the `<InputGroup>` component](https://github.com/patternfly/patternfly-react/pull/1108), the automatic assigning of `aria-describedby` by any non-form-control with an `id` was based off of that and an attempt to improve the a11y of that form field. Though I'd argue that's too arbitrary of a way to describe a form control, and usage of this component has changed in a way that this logic is too simple. From my perspective, the input group component is more of a layout that puts a group of "controls" (have the common border/box formatting - dropdown, select, text input, control buttons, etc) next to one another and otherwise isn't particularly opinionated about what goes inside of it.

**Comment by gitdallas** [comment]
@mcoker - i agree about inputGroup should not be so opinionated, and found it odd to have it automatically generate describedby on a button id. As far as disabling that.. what if you did have multiple elements with IDs and you only wanted a specific one that wasn't the first (right now it just finds the first)?

To me it would make more sense to just pass the value for describedby if you want it - but at this point that would be a breaking change. Perhaps we could add a formControlDescribedBy prop where if set it would use that and otherwise it could keep doing what it's doing? But then if you don't want a describedby even though you do want an ID on something within the inputgroup.. what then? empty string for formControlDescribedBy? allow formControlDescribedBy to be `string | false`?

As far as the NumberInput - It's creating an `inputGroup` just around the [minus][input][plus] with no way to even put a label within that inputgroup. Maybe something should change there?

**Comment by mcoker** [comment]
IMO, we shouldn't build on top of the functionality, just add the option to disable it. If we have the option to pass an `id` to the input group to describe a form control, it's no more work to just pass that to the form control itself, and a lot more straight forward. Also, there are more things that can go in input groups that presumably need good labels (and maybe description) just like those form controls - if we do this with form select, why not any other menu like that? Dropdown, select, a custom menu/menu-toggle, etc. Similar to a text input, we would describe text input group, search input, etc. Unless we build support for that, I think this "feature" is odd and not of that much value.

> As far as the NumberInput - It's creating an inputGroup just around the [minus][input][plus] with no way to even put a label within that inputgroup. Maybe something should change there?

If I understand the question, we have props for `aria-label` for all of the number input elements, and the ability to pass props directly to them (to add an `id` to the input to associate it with a form `<label>`'s `for` attribute or add `aria-describedby`, for example) so I think that's good?

https://github.com/patternfly/patternfly-react/blob/bb63f5f216f6582c1b8a07a06082c190c9ace34a/packages/react-core/src/components/NumberInput/NumberInput.tsx#L45-L55

**Comment by gitdallas** [comment]
@mcoker - this would override the describedby passed in `inputProps`... but I suppose if we add your `disableAssignDescribedby` or something it could be used together... or maybe just letting `inputProps.describedby` have priority would be enough

**Comment by mcarrano** [comment]
@nicolethoen @tlabaj wanted to get your thoughts on the above convo. Looks like the enhacement is working as expected from a UX perspective, but it's unclear to me whether we should try to address the issue that @gitdallas raises about IDs. Thoughts?

**Comment by gitdallas** [comment]
TLDR on coker and my exchange, our final options were narrowed down to:

1. allow `inputProps['aria-describedby']` to persist over the automatic describedby creation
 this would basically mean changing [this line in inputGroup](https://github.com/patternfly/patternfly-react/blob/bb63f5f216f6582c1b8a07a06082c190c9ace34a/packages/react-core/src/components/InputGroup/InputGroup.tsx#L35) to add ` && child.props['aria-describedby'] === undefined` ... the one caveat would be that if you didn't want any describedby at all, you could still put an empty string but you're just going to have an empty describedby property on the dom (does this matter)?
2. have a new property on `NumberInput` component that is something like `disabledAssignDescribedBy` that just bypasses the feature.. with this they could still add the inputProps describedby if they wanted and it would be added. if they didn't add an inputProps describedby, the describedby prop wouldn't exist at all

this whole thing seems pretty edge casey, but need to figure something out

**Comment by gitdallas** [comment]
I went ahead and just made it so that if someone adds aria-describedby to the input props, they don't get overridden by the existing feature. Seems least invasive for this edge case.

File: packages/react-core/src/components/NumberInput/examples/NumberInput.md [hunk 3]
```diff
@@ -57,6 +57,49 @@ class BasicNumberInput extends React.Component {
 }
 ```
 
+### Without onChange
+
+```js
+import React from 'react';
+import { NumberInput } from '@patternfly/react-core';
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/examples/NumberInput.md:64]: can this example live in separate file like the others?

File: packages/react-core/src/components/NumberInput/examples/NumberInput.md [hunk 4]
```diff
@@ -57,6 +57,49 @@ class BasicNumberInput extends React.Component {
 }
 ```
 
+### Without onChange
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/examples/NumberInput.md:60]: @mcoker @mcarrano should this example be added to core.

File: packages/react-core/src/components/NumberInput/NumberInput.tsx (hunk 1 above)
Comment (tlabaj) [reply packages/react-core/src/components/NumberInput/NumberInput.tsx]: @gitdallas that looks good tom me for the onchange.

File: packages/react-core/src/components/NumberInput/examples/NumberInput.md (hunk 4 above)
Comment (mcoker) [reply packages/react-core/src/components/NumberInput/examples/NumberInput.md:60]: I think so, though I wonder if we want to use a `readonly` input. Typically readonly implies the user can't modify it, but they still can here with the plus/minus buttons and using keyboard up/down arrows. For example, you can't use `readonly` with `required` since it's assumed the user can't interact with it.

Comment (gitdallas) [reply packages/react-core/src/components/NumberInput/examples/NumberInput.md:60]: I can remove, I just thought if we have the ability to modify with buttons only we should show it. lmk

File: packages/react-core/src/components/NumberInput/NumberInput.tsx (hunk 1 above)
Comment (nicolethoen) [reply packages/react-core/src/components/NumberInput/NumberInput.tsx]: +1

File: packages/react-core/src/components/NumberInput/examples/WithStatus.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import { NumberInput } from '@patternfly/react-core';
+
+export const WithStatus: React.FunctionComponent = () => {
+  const max = 10;
+  const min = 0;
+
+  const [validated, setValidated] = React.useState<'success' | 'error' | 'warning'>('success');
```
Comment (nicolethoen) [inline packages/react-core/src/components/NumberInput/examples/WithStatus.tsx:8]: I wonder if we should used an exported type here rather than expecting products to have to use this union explicitly in their typescript code.

File: packages/react-core/src/components/NumberInput/examples/WithStatus.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import { NumberInput, ValidatedOptions } from '@patternfly/react-core';
+
+export const WithStatus: React.FunctionComponent = () => {
+  const max = 10;
+  const min = 0;
+
+  const [validated, setValidated] = React.useState<ValidatedOptions>(ValidatedOptions.success);
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/examples/WithStatus.tsx:8]: Since you are using this `ValidatedOptions` examples, I would add it to the prop table (line 5 of the md file, `propComponents`) so consumers can easily see the definition.

**Review by tlabaj — CHANGES_REQUESTED** [review]
Looks great so far Dallas.  Just one little comment.

File: packages/react-core/src/components/NumberInput/examples/WithStatus.tsx (hunk 6 above)
Comment (gitdallas) [reply packages/react-core/src/components/NumberInput/examples/WithStatus.tsx:8]: I think `propComponents` is reserved for components while ValidatedOptions is just an enum. We could add it as the type for the `validated` property within NumberInput.. but I remember not doing that in the past for the reason of the parser that creates the documentation not being able to drill down. So within the NumberInput props, we'd see the validated prop like this:
image

Comment (gitdallas) [reply packages/react-core/src/components/NumberInput/examples/WithStatus.tsx:8]: or if we combine options... 
image

Comment (tlabaj) [reply packages/react-core/src/components/NumberInput/examples/WithStatus.tsx:8]: @gitdallas Yes you are right.   And that is why for our props we list about the values, for documentation.  I think your second screenshot here would be good.

**Review by tlabaj — APPROVED** [review]
Thanks Dallas.  LGTM

**Review by wise-king-sullyman — APPROVED** [review]
Looks perfect to me!

**Review by nicolethoen — COMMENTED** [review]
I think I need to do more research about `input type=number` because I'm surprised that the default screen reader experience doesn't seem to include updating the user to the current value of the input when using the VO+arrow keys to change the value of the input. Same is true when updating the value using the `+` and `-` buttons. I think someone using a screen reader may need a little more information. 
@thatblindgeye @jessiehuff do you have any more information about this scenario?

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 7]
```diff
@@ -16,6 +17,12 @@ export interface NumberInputProps extends React.HTMLProps<HTMLDivElement> {
   widthChars?: number;
   /** Indicates the whole number input should be disabled */
   isDisabled?: boolean;
+  /** Value to indicate if the input is modified to show that validation state.
+   * If set to warning, input will be modified to indicate warning state.
+   * If set to success, input will be modified to indicate valid state.
+   * If set to error,  input will be modified to indicate error state.
+   */
```
Comment (thatblindgeye) [inline packages/react-core/src/components/NumberInput/NumberInput.tsx:24]:
```suggestion
  /** Value to indicate the input's validation state when the input is modified. */
```
Not sure if we need the rest of the test i.e. "If set to...", but not a blocker if we keep it in.

File: packages/react-core/src/components/NumberInput/examples/WithStatus.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      setValidated(ValidatedOptions.error);
+    } else if (diff > 1) {
+      setValidated(ValidatedOptions.warning);
+    } else {
+      setValidated(ValidatedOptions.success);
+    }
+  };
+  return (
+    <>
+      Status will change to warning/error depending on how far from 5.
```
Comment (thatblindgeye) [inline packages/react-core/src/components/NumberInput/examples/WithStatus.tsx:41]: Nit: this could be reworded to something more along the lines of:

```suggestion
In the following example, the validated status will update based on the value of the number input and how far from the number 5 it is.
```

**Review by thatblindgeye — COMMENTED** [review]
This is looking good! I had mainly some nitpicks below (let me know what you think).

In addition to @nicolethoen  comment above, another issue is that the validation state only gets announced when navigating to the input, and even then only when the input is invalid (so a "warning" state won't get announced).

If the intent is to not have any validation text above/below the number input, a really quick option could be making the Number Input a live region and passing  `inputAriaLabel={validated}` instead of "number input" (the value of the input could also be passed in, so that both the validation state and value get announced which could resolve the issue Nicole mentioned above).

**Comment by thatblindgeye** [comment]
>do you have any more information about this scenario?

@nicolethoen it looks like this is partially an issue with Chromium (https://stackoverflow.com/questions/48299571/screen-reader-cannot-read-number-input-values-on-chrome). 

With Firefox, changing the Number Input via Voice Over + arrow keys while in the input has the new value announced, though pressing the - or + buttons does not announce the new value. Safari acts similary to Firefox with some slight differences (works better when you just focus the input and press up/down arrow rather than "entering" the input via VO modifier + Shift + Down).

It is definitely odd that the change in value doesn't get announced in Chrome by default, even when changing the value via the actual input and not our own +/- buttons.

Adding aria live like I noted above seems to help resolve that in Chrome, but in Firefox the value ends up getting repeated by VO (pressing the minus button once would announce "89, 89, 89, 89" for me). So *that* may require a little more investigation.

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 9]
```diff
@@ -16,6 +17,12 @@ export interface NumberInputProps extends React.HTMLProps<HTMLDivElement> {
...
+  validated?: 'default' | 'error' | 'warning' | 'success' | ValidatedOptions;
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/NumberInput.tsx:25]: Can we actually mark this prop beta please.

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 10]
```diff
@@ -122,10 +132,15 @@ export const NumberInput: React.FunctionComponent<NumberInputProps> = ({
           </span>
         </Button>
         <input
-          className={css(styles.formControl)}
+          className={css(
+            styles.formControl,
+            validated === ValidatedOptions.success && formControlStyles.modifiers.success,
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/NumberInput.tsx]: @mcoker Should these `formControlStyles` be available in the number input.  If so, should we open a core issue?

**Review by thatblindgeye — APPROVED** [review]
Looks good! Opened #7871 as a followup to the a11y issues since it looks like that will require more investigation to find a solution that resolves Chromium issues without introducing issues in other browsers.
